ENH: adding multi-target dijsktra performance enhancement

fcollman opened the pull request:
for some applications one wants to find the shortest path to any target node.  This pull request adds an optional flag to the dijsktra csgraph function to search for the shortest path to ANY of the set of indices passed, as opposed to the current implementation which always calculates the shortest path to EVERY index.   In practice with large graphs this can be a very large difference in performance, on my machine with my problem it was 75x improvement in speed.  Some applications want one of these, others want the other so I thought a flag was the most appropriate implementation, though a separately named call might also be appropriate.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -428,8 +428,13 @@ def dijkstra(csgraph, directed=True, indices=None,
+        If True, find the shortest path to any index in indices,
+        resulting distance matrix with be Kx1 vector, where K is the len(indices)
+        and predecessor matrix will be Nx1. Much faster if this is what you want.
+        .. versionadded:: 0.15.0
```
A few typos in this doc block, and also the `versionadded` was for another argument so shouldn't have been modified (and we aren't on version `0.15.0` either). If the feature does get added you'd probably want a new `versionadded` for i.e., `1.3.0`.

fcollman replied:
I think i've addressed the typos in the latest changes, and altered the version to 1.3.0. Please let me know wha the accepted protocol for that is.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -614,6 +645,70 @@ cdef _dijkstra_directed(
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indices,
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indptr,
+            np.ndarray[DTYPE_t, ndim=1, mode='c'] dist_matrix,
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] pred,
```
I suspect there's probably a preference for typed memoryviews in new Cython code, although you'd be correct in noticing that there are cases of the older buffer usage in this module

fcollman replied:
I'll work on changing this code style tomorrow and post a note when I think its done.

fcollman replied:
I believe I have integrated the memory views, although line 344 errored out when i changed the signature of _floyd_warshall, so I have left that function in the old style.  I'm not experienced enough with memoryarray's to be sure why.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -614,6 +645,70 @@ cdef _dijkstra_directed(
+    cdef DTYPE_t next_val
+
+    cdef int return_pred = (pred.size > 0)
+
```
There are so many `cdef` lines that putting them all in a `cdef:` block might be preferable. Also, not sure if the occasional blank lines between `cdef`s are helpful or not.

fcollman replied:
I submitted a commit that changes this style across the entire module, could be considered a different PR, but let me know what the standards are here.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -696,6 +791,87 @@ cdef _dijkstra_undirected(
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] csrT_indices,
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] csrT_indptr,
+            np.ndarray[DTYPE_t, ndim=1, mode='c'] dist_matrix,
+            np.ndarray[ITYPE_t, ndim=1, mode='c'] pred,
```
probably want typed memoryviews as noted above

fcollman replied:
this is now addressed I believed.

fcollman commented:
The failed tests appear to be related to the bug being addressed by #9662

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -428,8 +428,12 @@ def dijkstra(csgraph, directed=True, indices=None,
         that are separated by a distance > limit. For such pairs, the distance
         will be equal to np.inf (i.e., not connected).
-
-        .. versionadded:: 0.14.0
```
still missing the old `versionadded`, which is for another argument -- I think that should just stay where it was

fcollman replied:
ah i see.. fixed

tylerjereddy commented on scipy/sparse/csgraph/tests/test_shortest_path.py:
```diff
@@ -96,6 +96,15 @@ def check(method, directed_in):
 
+def test_dijkstra_indices_multi_target():
+    indices = np.arange(4)
+    SP = dijkstra(directed_G, directed=False, indices= indices, multi_target=True)
```
extra space at `indices` can be removed I think

fcollman replied:
done and other flake8 fixes

fcollman commented:
multiple target and/or source djisktra is a feature of other graph libraries
see for instances...

https://igraph.org/python/doc/igraph.GraphBase-class.html#shortest_paths

csgraph doesn't have a notion of both sources and targets in its vertex filtering, so this is the best i can do without adding much more to the api.

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -428,8 +428,13 @@ def dijkstra(csgraph, directed=True, indices=None,
+    multi_target: bool, optional
+        If False (default), independently find the shortest path to every index
+        in indices.  If True, find the shortest path to any index in indices and
+        the resulting distance matrix with be Kx1 vector, where K is the len(indices)
```
### Readability

I find these docs a bit difficult to decode. How about:
```
    multi_target: bool, optional
￼        If False (default), for every node in the graph, find the shortest path to every node in indices.
        If True, for every node in the graph, find the shortest path to any of the nodes in indices (which
        can be substantially faster).
...

    dist_matrix : ndarray, shape ([n_indices, ]n_nodes,)
        The matrix of distances between graph nodes. If multi_target=False, dist_matrix
        has shape (n_indices, n_nodes) and dist_matrix[i, j] gives the shortest distance
        from point i to point j along the graph. If mult_target=True, dist_matrix has
        shape (n_nodes,) and contains the shortest path from each node to any
        of the nodes in indices.
    predecessors : ndarray, shape ([n_indices, ]n_nodes,)
        If multi_target=False, this has shape (n_indices, n_nodes),
        otherwise it has shape (n_nodes,).
        ...
```
... tweaked to be PEP8 compatible, of course.

### Completeness

One thing I find missing in this implementation is an output of the set of indices that yielded the minimal distances. Maybe this should be added to the end of `Returns` and returned as:
```
    min_indices : ndarray, shape (n_nodes,)
        Only returned if multi_target=True. Contains the node numbers
        that yielded the minimum distances.
```
In other words, in many cases you may want to know which nodes gave you the min distances, and there should be almost no cost to outputting this value as well as the distances themselves.

### Naming

I don't find `multi_target` to be the clearest naming. What it seems to be doing is more a minimum across all `indices`, so a name like `min_only=True` to compute/return the minimum of the shortest distances probably makes more sense than `multi_target=True`.

fcollman replied:
I like the doc changes, i've incorporated them locally, will push after i do tests.  I also like min_only better than multi_target, so i've adopted that. 

In terms of the output, this would require some deeper modification of the C components in the Fibonacci heap to allow for tracking which target node was the reason for each of shortest paths on each source node reached.  I think this is an extra optimization, as the information is embedded in the pred network and comes with an added memory cost.  I could implement the 'expensive' version of this easily by just taking on a loop that follows along the pred path after getting them.  I've added that expensive version as a test now.

larsoner replied:
Sounds good, does it hurt performance much?

fcollman replied:
so for any particular node, you start at that node's index in pred, and then you go to the index at that spot, hopping your way along till you encounter a -9999 entry.  You should only get there if it's impossible to reach a target (in which case you make 0 hops), or when you end up on one of the target indices you passed in (in which you make a number of hops equal to the path from your starting node to the closest target).

fcollman replied:
my new test actually implements this logic now to do the check you suggested.

larsoner replied:
It makes sense this would be true for the output you get in the `indices=None, min_index=False` case, but it doesn't look like the `min_index=True` case has a `predecessors` matrix of the correct shape to traverse this way. Am I missing something?

larsoner replied:
Okay if the tests pass (which it looks like they did) I must just be missing something about how `pred` works. So I'll trust the result there :)

fcollman replied:
i have now implemented this tracking so you get back what index was closest to each target when doing min_only.

larsoner commented on scipy/sparse/csgraph/tests/test_shortest_path.py:
```diff
@@ -97,6 +97,22 @@ def check(method, directed_in):
+                  directed=True,
+                  indices=indices,
+                  multi_target=True)
+    assert_array_almost_equal(SP, ans)
```
If you add the `min_indices` output proposed above, you can also check that the `argmax` of the `multi_target=False` (or maybe now `min_only=False`) result is the same as the `min_indices` of the `multi_target/min_only=True` case.

fcollman replied:
see above comment.

fcollman replied:
see above follow on comment... i have now fixed the tests to use this 'sources' output to check against the answers, as well as fixing the tests to use more combinations of indices for testing.

fcollman commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,41 @@
+try:
+    from scipy.sparse.csgraph import dijkstra
+except ImportError:
+    pass
```
i was just following the instructions here... https://github.com/scipy/scipy/tree/master/benchmarks#writing-benchmarks

fcollman commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,41 @@
+    param_names = ['n', 'format', 'normed']
+
+    def setup(self, n):
+        # make a random connectivity matrix
```
i had this at first and then removed it because I saw it was here.. 
https://github.com/scipy/scipy/blob/master/benchmarks/benchmarks/__init__.py

fcollman replied:
yeah i have to say i agree with you, but I was just trying to adopt to the broader conventions of the benchmark suite.

larsoner replied:
Okay we can fix it globally later at some point rather than diverge here

fcollman commented:
i rebased off master to get the test fixes from #9662

fcollman commented:
I think is now ready for review again.. I have also consolidated many of the similar code blocks into subfunctions, rather than having lots of repeated code.  There is likely more one could do, but it's better than before.

larsoner commented:
> I have also consolidated many of the similar code blocks into subfunctions, rather than having lots of repeated code.

Excellent, I noticed when looking that it seemed like there was some redundancy, but wasn't sure how simple the refactoring would be. Glad you have worked it out

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -430,22 +430,39 @@ def dijkstra(csgraph, directed=True, indices=None,
+    min_only: bool, optional
+￼        If False (default), for every node in the graph, find the shortest path to every node in indices.
+        If True, for every node in the graph, find the shortest path to any of the nodes in indices (which
+        can be substantially faster).
```
Can you PEP8-ify these docstrings (<80 char)?

fcollman replied:
done

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -430,22 +430,39 @@ def dijkstra(csgraph, directed=True, indices=None,
         path from point i to point j.  If no path exists between point
         i and j, then predecessors[i, j] = -9999
-
+    sources: ndarray, shape(n_nodes)
```
shape (n_nodes,)

fcollman replied:
done

fcollman replied:
done

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -430,22 +430,39 @@ def dijkstra(csgraph, directed=True, indices=None,
         i and j, then predecessors[i, j] = -9999
-
+    sources: ndarray, shape(n_nodes)
+        only returned if min_only=True and return_predecessors=True.
```
Needs proper capitalization on start of sentences

fcollman replied:
done

larsoner commented on scipy/sparse/csgraph/tests/test_shortest_path.py:
```diff
@@ -97,6 +97,34 @@ def check(method, directed_in):
+        assert_array_equal(min_ind_ans, sources)
+    for indices in index_list:
+        for directed in (True, False):
+            check(directed, indices)
```
This formulation could be streamlined and rendered more readable as:
```
@pytest.mark.parametrize('directed, SP_ans',
                         ((True, directed_SP),
                          (False, undirected_SP))
@pytest.mark.parametrize('indices', ([0, 2, 4], [0, 4], [3, 4])
def test_dijkstra_indices_min_only(directed, SP_ans, indices):
    <contents of check function>
```

larsoner replied:
You can. We are adopting `pytest` practices bit by bit. Might as well take advantage of the readability improvements they would provide in this PR

fcollman replied:
I can leave it half done you mean?

larsoner replied:
Yes. It's possible that we will (effectively) never bother to change the existing tests to `pytest` style, since they work fine and doing so makes things like `git blame` and history browsing harder anyway.

fcollman replied:
push a commit with the parameterized tests

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -655,47 +841,98 @@ cdef _dijkstra_undirected(
+        dist_matrix[v.index] = v.val
+
+    free(nodes)
+
```
PEP8 wants exactly two newlines between root-level function defs

fcollman replied:
fixed

fcollman commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -1326,30 +1563,32 @@ cdef void insert_node(FibonacciHeap* heap,
-    cdef FibonacciNode *linknode
-    cdef FibonacciNode *parent
-    cdef FibonacciNode *child
+    cdef FibonacciNode * linknode
```
i'm sorry about this... i didn't realize that these autopep8 changes would be so dramatic.. i'll revert anf fix it.

fcollman replied:
i've addressed this

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -655,47 +841,98 @@ cdef _dijkstra_undirected(
+                                  pred,
+                                  sources,
+                                  return_pred,
+                                  limit)
```
I think this long-list format actually hurts readability a bit, as it makes the file much longer than necessary. WDYT about doing:
```
        _dijkstra_scan_heap_multi(
            &heap, v, nodes, csr_weights, csr_indices, csr_indptr,
            pred, sources, return_pred, limit)
```

fcollman replied:
agreed.. i tried to consolidate related variables on the same line, slightly different than your suggestion but i hope more readable.

larsoner commented:
Code looks reasonable to me. And the tests look much nicer now :)

@tylerjereddy will you be able to look at the Cython changes in a bit more detail? It would be good to get another set of eyes on it.

fcollman commented:
Some of the failing tests seem unrelated, but there are some documentation failures and unicode errors in the travis-ci run that I don't understand where they are coming from. If someone could help me understand what's wrong I'm happy to fix it.

larsoner commented:
The error says you have added unicode chars without declaring the encoding at the top of the file. Specifically:
```
larsoner@bunk:~/python/scipy$ grep -rIP '[^\x00-\x7F]' scipy/sparse/csgraph/
scipy/sparse/csgraph/_shortest_path.pyx:￼        If False (default), for every node in the graph, find the shortest path
```
It looks like if I replace the whitespace at the beginning of that line with spaces it works. So somehow you had a unicode whitespace character. I've pushed a commit to fix it. (Feel free to force push right over it if you'd rather fix it or don't want to `pull`, etc.)

larsoner commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -63,9 +63,9 @@ def shortest_path(csgraph, method='auto',
-		     representation.
+            ``k`` is the average number of connected edges per node.
+            The input csgraph will be converted to a csr
+            representation.
```
This was also causing a failure

perimosocordiae commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +535,198 @@ def dijkstra(csgraph, directed=True, indices=None,
-    dist_matrix.fill(np.inf)
-    dist_matrix[np.arange(len(indices)), indices] = 0
+    if min_only:
+        dist_matrix = np.zeros((N), dtype=DTYPE)
```
`dist_matrix = np.full(N, np.inf, dtype=DTYPE)` would be simpler here, and in the other cases below.

perimosocordiae commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +535,198 @@ def dijkstra(csgraph, directed=True, indices=None,
-        predecessor_matrix = np.empty((0, N), dtype=ITYPE)
+        if min_only:
+            predecessor_matrix = np.empty((0), dtype=ITYPE)
+            source_matrix = np.empty((0), dtype=ITYPE)
```
The extra parens aren't necessary. If you want it to be a one-element tuple, you'll need to add a comma: `(0,)`.

But I'd go with the plain `0`, personally.

perimosocordiae commented on scipy/sparse/csgraph/tests/test_shortest_path.py:
```diff
@@ -97,6 +98,33 @@ def check(method, directed_in):
+                                 return_predecessors=True)
+    assert_array_almost_equal(SP, min_d_ans)
+    assert_array_equal(min_ind_ans, sources)
+    SP = dijkstra(directed_G,
```
Why aren't we testing the result of this call?

fcollman commented:
In using this myself I found a bug with the new feature that only presented itself with non uniform edge weights.. i've added a test for source consistency and fixed that bug.  However, I'm wondering what the process is to move this PR forward, I would be happy to do any work needed but I'm still a bit unclear on how the enhancement addition process is managed.  I think I've addressed all the points raised by the reviews so far.  @larsoner  @tylerjereddy @perimosocordiae

tylerjereddy commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,41 @@
+                      directed=False,
+                      indices=self.indices,
+                      min_only=False)
+        ds = np.min(dm, axis=1) 
```
I'm not sure if we want this extra calculation in here. I think I understand why it is there now -- to objectively compare performance with the `min_only=True` version, but on the other hand we're now benchmarking a NumPy function and a SciPy function in the same `time_` function.

If we do remove that final step, they can be combined to a single parametrized bench.

fcollman replied:
understood... i have removed the np.min benchmark.

tylerjereddy commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,41 @@
+    param_names = ['n', 'format', 'normed']
+
+    def setup(self, n):
+        # make a random connectivity matrix
```
Quite a few of our benchmarking modules do manually set the random seed. Are we confident that the `__init__` file is actually always going to do that? What about the case where `asv` is used directly instead of our in-house `run.py`?

fcollman replied:
added a random seed setting

tylerjereddy commented on scipy/sparse/csgraph/tests/test_shortest_path.py:
```diff
@@ -97,6 +99,56 @@ def check(method, directed_in):
+    data.setdiag(np.zeros(n, dtype=np.bool))
+    # choose some random vertices
+    v = np.arange(n)
+    np.random.shuffle(v)
```
do we need to pin the seed here?

fcollman replied:
done

fcollman commented:
> What happens if there are two or more paths between nodes with identical distances? Does the algorithm consistently return the same path (i.e., "stable"), or can it sometimes pick other paths of identical travel distance? Do we have a unit test for that?

That's a good question.  The algorithm is doing a  breadth first search, starting from the first index passed, and it will only update the shortest distance/path to a node if it finds a shorter path later in the search, so who gets there first matters.  However, I don't see any non deterministic aspects of the codebase, so the order of how the starting indices are presented would affect the result, but the same order should result in the same result.  I can add an explicit unit test running things many times, but I'm not sure that's the extent of the worry you had as different machines or apparently similar versions of the same problem might result in different answers.  Say for example if you passed it a view of a numpy array of indices, versus an continguousarray of the same thing.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
+        if min_only:
+            predecessor_matrix = np.empty((N), dtype=ITYPE)
+            predecessor_matrix.fill(NULL_IDX)
+            source_matrix = np.empty((N), dtype=ITYPE)
```
there may be some benefit to typing `source matrix`, but it is also true that the original Cython function wasn't thoroughly typed, so that job could perhaps fall to a future PR

fcollman replied:
I'm not familiar enough with ctyping to know if the other reshape operations and things that are needed in this python function to return things in the appropriate python format are compatible with this.. i think we should leave this to another PR as you suggested.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
-    cdef unsigned int N = dist_matrix.shape[1]
-    cdef unsigned int i, k, j_source, j_current
-    cdef ITYPE_t j
+cdef _dijkstra_setup_heap_multi(FibonacciHeap *heap,
```
There may be some benefit to decorating with `@cython.boundscheck(False)` based on the cython html report, but you'll know better than I if this is safe. I guess this function should only get called from C, so maybe it would be safe?

fcollman replied:
I agree this is only called from C and so should be safe.  Added the bounds decorator.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
+        current_node.source = j_source
+        insert_node(heap, &nodes[j_source])
+
+cdef _dijkstra_scan_heap_multi(FibonacciHeap *heap,
```
same here if bounds checking can safely be deactivated

fcollman replied:
done.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
+                        pred[j_current] = v.index
+                        sources[j_current] = v.source
+
+cdef _dijkstra_scan_heap(FibonacciHeap *heap,
```
and here if safe to deactivate bounds checking

fcollman replied:
done.

tylerjereddy commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
-    cdef FibonacciNode *current_node
-    cdef FibonacciNode* nodes = <FibonacciNode*> malloc(N *
-                                                        sizeof(FibonacciNode))
+cdef _dijkstra_directed(
```
likewise on bounds checking, I think

fcollman replied:
done.

fcollman commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,35 @@
+    params = [
+        [30, 300, 900]
+    ]
+    param_names = ['n', 'format', 'normed']
```
you are right.. i copied that and forgot to remove the unused ones.  I have adding min_only as true false and removed the others.

fcollman commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,35 @@
+        return dijkstra(self.data,
+                        directed=False,
+                        indices=self.indices,
+                        min_only=True)
```
done.. see above comment.

fcollman commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -391,10 +390,10 @@ cdef void _floyd_warshall(
              return_predecessors=False,
-             unweighted=False, limit=np.inf):
+             unweighted=False, limit=np.inf,
+             min_only=False):
```
done.

fcollman commented:
I have rebased off master, please advise as to how commits should be cleaned up.  a) adding the BUG, ENH, FIX type tags.  b) doing a major squash in order to condense them to a much smaller set (and if so under what major feature steps do you see). I could see.. 1) adding first min_only feature. 2) adding source support. 3) adding tests. 4) adding benchmarks. 5) documentation 6) refactoring duplicate code in module 7) fixing bugs.

fcollman commented:
btw the benchmarks now still show a nice 9x effect on the min_only feature, the min calculation of course adds to this significantly, but I see the reasoning for taking that out.
``` ===== ============= ============
              --             min_only
              ----- --------------------------
                n        True        False
              ===== ============= ============
                30     585±20μs     608±10μs
               300   4.77±0.02ms   18.8±0.2ms
               900    42.3±0.6ms    373±10ms
              ===== ============= ============```

fcollman commented:
seems the lil format was much worse in performance difference between the options.. reverting to csc gets back a slightly improved performance from what i had originally 
```===== ============= ============
              --             min_only
              ----- --------------------------
                n        True        False
              ===== ============= ============
                30     462±20μs     517±4μs
               300   1.11±0.01ms   15.9±0.4ms
               900    6.64±0.1ms    346±10ms
              ===== ============= ============```

fcollman commented:
@tylerjereddy would appreciate any advice on how you would like the commits squashed.

fcollman commented:
i've tried a first pass to get it down to 6 commits.

tylerjereddy commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,37 @@
+    def setup(self, n, min_only):
+        np.random.seed(1234)
+        # make a random connectivity matrix
+        data = scipy.sparse.rand(n, n, density=0.2, format='csc', random_state=42, dtype=np.bool)
```
this line is a bit long at 97 characters

tylerjereddy commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,37 @@
+        v = np.arange(n)
+        np.random.shuffle(v)
+        self.indices = v[:int(n*.1)]
+        self.min_only = min_only
```
this may not be needed since `min_only` is fed in below as a parameter anyway?

tylerjereddy commented on benchmarks/benchmarks/sparse_csgraph_djisktra.py:
```diff
@@ -0,0 +1,37 @@
+        self.min_only = min_only
+
+    def time_dijkstra_multi(self, n, min_only):
+        return dijkstra(self.data,
```
probably don't need the `return`

tylerjereddy commented:
I addressed my own comments & rebased / resolved merge conflicts.

I'll aim to merge if CI is still all green.

tylerjereddy commented:
All green so in it goes, thanks @fcollman. Adding a suitable release note entry would be helpful: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0

fcollman commented:
awesome.. thank you @tylerjereddy  I have added a note to the wiki.
